content/install-hook-userstylesworld.js

```diff
@@ -2,48 +2,42 @@
 'use strict';
 
 (() => {
-  const allowedOrigin = 'https://userstyles.world';
+  const ORIGIN = 'https://userstyles.world';
+  const HANDLERS = Object.assign(Object.create(null), {
 
-  const sendPostMessage = message => {
-    if (allowedOrigin === location.origin) {
-      window.postMessage(message, location.origin);
-    }
-  };
-
-  const onPageLoaded = event => {
-    if (event.data
-    && allowedOrigin === event.origin
-    ) {
-      switch (event.data.type) {
-        case 'usw-ready': {
-          sendPostMessage({type: 'usw-remove-stylus-button'});
+    async 'usw-ready'() {
+      send({type: 'usw-remove-stylus-button'});
+      if (location.pathname === '/api/oauth/style/new') {
+        const styleId = Number(new URLSearchParams(location.search).get('vendor_data'));
+        const data = await API.data.pop('usw' + styleId);
+        send({type: 'usw-fill-new-style', data});
+      }
+    },
 
-          if (location.pathname === '/api/oauth/style/new') {
-            const styleId = Number(new URLSearchParams(location.search).get('vendor_data'));
-            API.data.pop('usw' + styleId).then(data => {
-              sendPostMessage({type: 'usw-fill-new-style', data});
-            });
-          }
-          break;
-        }
-        case 'usw-style-info-request': {
-          switch (event.data.requestType) {
-            case 'installed': {
-              API.styles.find({updateUrl: `https://userstyles.world/api/style/${event.data.styleID}.user.css`})
-                .then(style => {
-                  sendPostMessage({
-                    type: 'usw-style-info-response',
-                    data: {installed: Boolean(style), requestType: 'installed'},
-                  });
-                });
-              break;
-            }
-          }
+    async 'usw-style-info-request'(data) {
+      switch (data.requestType) {
+        case 'installed': {
+          const updateUrl = `${ORIGIN}/api/style/${data.styleID}.user.css`;
+          const style = await API.styles.find({updateUrl});
+          send({
+            type: 'usw-style-info-response',
+            data: {installed: Boolean(style), requestType: 'installed'},
+          });
           break;
         }
       }
+    },
+  });
+
+  window.addEventListener('message', ({data, source}) => {
+    // Accepting events only from this page.
+    if (data && source === window) {
```

Comment:
![image](https://user-images.githubusercontent.com/25481501/127660186-71f0937f-7621-455b-a3ee-0ab220ba6314.png)
![image](https://user-images.githubusercontent.com/25481501/127660320-4d7194a4-b975-44c6-9f09-2cea8609614b.png)

Both in chromium and firefox "source" seems to be null. I'm not sure if we need to go pass this manually. But it seems a better or alternative choice is `target` or any other property which return `Window`

Reply:
Thanks for catching, it was really bad.

Indeed, MDN says Firefox doesn't reveal `source` to extensions for "security" reasons, which sounds to me like they actually couldn't implement it due to the specifics of their implementation of content scripts.

But `target` won't help as it will always be the target window even if the message was sent from another window.

I guess I'll add a check for FF then.